python/lsst/ts/observatory/control/auxtel/atcs.py

```diff
@@ -1524,6 +1524,140 @@ def unset_azel_slew_checks(self, checks):
+
+        Raise
+        -----
+        RuntimeError:
+            If the last `atspectrographCorrectionCompleted` event was published
+            before the last `atspectrographCorrectionStarted` and a timeout
+            happens when waiting for a new `atspectrographCorrectionCompleted`.
+        """
+
+        ataos_enabled_corrections = await self.rem.ataos.evt_correctionEnabled.aget(
+            timeout=self.fast_timeout
+        )
+        # If the atspectrograph correction is not enabled, then nothing to wait
+        # for. Just return.
+        if not ataos_enabled_corrections.atspectrograph:
+            return True
+
+        self.log.debug(
+            "atspectrograph correction running. Trying to determine state of the corrections."
+        )
+        ret_val = await asyncio.gather(
+            self.rem.ataos.evt_atspectrographCorrectionStarted.next(
+                flush=True, timeout=self.fast_timeout
+            ),
+            self.rem.ataos.evt_atspectrographCorrectionCompleted.next(
+                flush=True, timeout=self.fast_timeout
+            ),
+            return_exceptions=True,
+        )
+        # If they are both execeptions it probably means the ATAOS got into a
+        # stable state and it is ready to take images. We need to verify now
+        # that the last completed correction was issued after the last started
+        # correction.
+        if all([isinstance(val, asyncio.TimeoutError) for val in ret_val]):
+            self.log.debug(
+                f"No correction seen in the last {self.fast_timeout} seconds. "
+                "Determining order of last corrections."
+            )
+            started, completed = await asyncio.gather(
+                self.rem.ataos.evt_atspectrographCorrectionStarted.aget(
+                    timeout=self.fast_timeout
+                ),
+                self.rem.ataos.evt_atspectrographCorrectionCompleted.aget(
+                    timeout=self.fast_timeout
+                ),
+            )
+            if started.private_sndStamp >= completed.private_sndStamp:
+                self.log.debug("Last correction still pending.")
+                # There is probably a correction going on. Still not ready.
+                return False
+            else:
+                self.log.debug("Last correction completed.")
+                # Correction completed was done after last started. It is
+                # probably ready to take data.
+                return True
+
+        else:
+            # This means we got some data. We are not in a stable state yet.
+            correction_name = ["started", "completed"]
+            correction_received = [
+                name
+                for name, corr in zip(correction_name, ret_val)
+                if not isinstance(corr, asyncio.TimeoutError)
+            ]
+
+            self.log.debug(f"Received corrections: {correction_received}.")
+            return False
```

Comment:
My main concern with this approach is that `ret_val` may simply have retrieved the latest CorrectionStarted and CorrectionCompleted events so technically speaking the corrections have been applied but the code forces to wait for the next invocation where those two events time out. This seems like unnecessarily waiting for a timeout to happen. If `ret_val` contains data in the outer else block then applying the corrections may have finished if `private_sndStamp` of CorrectionStarted < `private_sndStamp` of CorrectionCompleted, right?

Reply:
so, the problem is that, if a correction was completed you cannot guarantee that there won't be another one coming just after it. The only way to be more certain about it is to make sure all corrections scheduled where completed and no new correction is done in that grace period of time.

Reply:
OK, thanks for the explanation.